django/db/models/sql/compiler.py

```diff
@@ -906,14 +906,17 @@ def local_setter(obj, from_obj):
                         if from_obj:
                             f.remote_field.set_cached_value(from_obj, obj)
 
-                    def remote_setter(obj, from_obj):
-                        setattr(from_obj, name, obj)
+                    def remote_setter(name):
+                        def inner(obj, from_obj):
+                            setattr(from_obj, name, obj)
+                        return inner
+
                     klass_info = {
                         'model': model,
                         'field': f,
                         'reverse': True,
                         'local_setter': local_setter,
-                        'remote_setter': remote_setter,
+                        'remote_setter': remote_setter(name),
```

Comment:
Both `local_setter` and `remote_setter` could be defined outside the loop and `remote_setter` could be closed over `name` with a `partial`

e.g.

```python
def local_setter(obj, from_obj):
   ...

def remote_setter(name, obj, from_obj): 
    setattr(from_obj, name, obj)

for name in list(requested):
```

```suggestion
                        'remote_setter': partial(remote_setter, name=name)
```

Reply:
@charettes Thanks for the comment. :+1: for `partial`
@felixxm Thanks for the test!